File: estrutura-de-dados-avancada/abp.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
  int v;
  struct Node *l;
  struct Node *r;
} Node;

Node *create_node(int v) {
  Node *n = malloc(sizeof(Node));
  n->v = v;
  n->l = NULL;
  n->r = NULL;

  return n;
}

void pre_order(Node *tree) {
  if (tree == NULL) {
    return;
  }
  printf("%d ", tree->v);
  pre_order(tree->l);
  pre_order(tree->r);
}

void in_order(Node *tree) {
  if (tree == NULL) {
    return;
  }
  pre_order(tree->l);
  printf("%d ", tree->v);
  pre_order(tree->r);
}

void post_order(Node *tree) {
  if (tree == NULL) {
    return;
  }
  pre_order(tree->l);
  pre_order(tree->r);
  printf("%d ", tree->v);
}

Node *add_rec(Node *node, int i) {
  if (node == NULL) {
    return create_node(i);
  }

  if (i < node->v) {
    node->l = add_rec(node->l, i);
  } else if (i > node->v) {
    node->r = add_rec(node->r, i);
  }

  return node;
}

void add(Node *node, int i) { node = add_rec(node, i); }

Node *search(Node *node, int v) {
  if (node == NULL || node->v == v) {
    return node;
  }
  if (v < node->v) {
    return search(node->l, v);
  }
  return search(node->r, v);
}

int main(void) {
  Node *tree = create_node(10);
  add(tree, 12);
  add(tree, 5);
  add(tree, 2);
  add(tree, 7);

  Node *r = search(NULL, 1);
  if (r != NULL) {
    printf("Encontrado = %d\n", r->v);
  }

  printf("pre order\t[");
  pre_order(tree);
  printf("]\n");

  printf("in order\t[");
  in_order(tree);
  printf("]\n");

  printf("post order\t[");
  post_order(tree);
  printf("]\n");

  return 0;
}

File: estrutura-de-dados-3/abp.go
package main

import "fmt"

type Node struct {
	value int
	left  *Node
	right *Node
}

func create(v int) *Node {
	return &Node{
		value: v,
	}
}

func (n *Node) pre_order() {
	if n == nil {
		return
	}
	fmt.Printf("%d ", n.value)
	n.left.pre_order()
	n.right.pre_order()
}

func (n *Node) addRec(v int) *Node {
	if n == nil {
		return create(v)
	}
	if v < n.value {
		n.left = n.left.addRec(v)
	} else if v > n.value {
		n.right = n.right.addRec(v)
	}
	return n
}

func (n *Node) add(v int) {
	n = n.addRec(v)
}

func (n *Node) serach(v int) *Node {
	if n == nil || n.value == v {
		return n
	}
	if v < n.value {
		return n.left.serach(v)
	}
	return n.right.serach(v)
}

func main() {
	tree := create(10)
	tree.add(5)
	tree.add(2)
	tree.add(7)
	tree.add(12)
	tree.pre_order()
	println()

	if r := tree.serach(10); r != nil {
		fmt.Printf("encontrou %d\n", r.value)
	}
}
